R/normalize.R

```diff
@@ -0,0 +1,175 @@
+      new_comps <- head(new_comps, -1)
+    }
+  }
+
+  res <- paste(new_comps, collapse = "/")
+  if (initial_slashes) {
+    res <- paste0(c(rep("/", initial_slashes), res), collapse = "")
+  }
+
+  if (res != "") res else "."
+}
+
+
+win_path_norm1 <- function(path1) {
+
+  ## in the case of paths with these prefixes:
+  ## \\.\ -> device names
+  ## \\?\ -> literal paths
+  ## do not do any normalization, but return the path unchanged
+  if (startswith(path1, "\\\\.\\") ||
+      startswith(path1, "\\\\?\\")) return(path1)
+
+  path1 <- gsub("/", "\\", path1, fixed = TRUE)
+  spl <- path_split_drive1(path1)
+  prefix <- spl[1]
+  path1 <- spl[2]
+
+  ## Collapse initial backslashes
+  if (startswith(path1, "\\")) {
+    prefix <- paste0(prefix, "\\")
+    path1 <- substr(path1, 2, nchar(path1))
+  }
+
+  comps <- as.list(strsplit(path1, "\\", fixed = TRUE)[[1]])
+
+  i <- 1
+  while (i <= length(comps)) {
+
+    if (comps[[i]] == "" || comps[[i]] == ".") {
+      comps[[i]] <- NULL
+
+    } else if (comps[[i]] == win_pardir) {
+      if (i > 1 && comps[[i - 1]] != win_pardir) {
+        comps[[i - 1]] <- NULL
+        comps[[i - 1]] <- NULL
+        i <- i - 1
+      } else if (i == 1 && endswith(prefix, "\\")) {
+        comps[[i]] <- NULL
+      } else {
+        i <- i + 1
+      }
+
+    } else {
+      i <- i + 1
+    }
+  }
+
+  if (prefix == "" && length(comps) == 0) comps <- "."
+
+  paste0(prefix, paste(unlist(comps), collapse = "\\"))
+}
+
+##' Split a pathname into drive/UNC sharepoint and relative path specifiers
+##'
+##' Returns a list with 2 element character vectors: drive/UNC and path,
+##' either part may be empty string.
+##'
+##' It always holds that \code{paste0(output[1], output[2]) == input}.
+##'
+##' If the path contained a drive letter, the drive/UNC part will contain
+##' everything up to and including the colon. E.g.
+##' \code{path_split_drive("c:/dir")} returns \code{list(c("c:", "/dir"))}.
+##'
+##' If the path contained a UNC path, the drive/UNC will contain the host
+##' name and share up to but not including the fourth directory separator
+##' character. E.g. \code{path_split_drive("//host/computer/dir")} returns
+##' \code{list(c("//host/computer", "/dir"))}.
+##'
+##' Paths cannot contain both a drive letter and a UNC path.
+##'
+##' @param path Character vector.
+##' @return List of character vectors of length two. The first elements
+##'   contain the drive letters or UNC paths, the second elements the
+##'   relative paths.
+##'
+##' @export
+##' @examples
+##' path_split_drive(c("c:/Rtools/bin", "\\\\machine\\share\\folder"))
+
+path_split_drive <- function(path) {
+  lapply(path, path_split_drive1)
+}
+
+path_split_drive1 <- function(path1) {
+  npath1 <- gsub("/", "\\", path1, fixed = TRUE)
+  if (startswith(npath1, "\\\\") && substr(npath1, 3, 3) != "\\") {
+    if (grepl("^\\\\\\\\[^\\\\]+\\\\[^\\\\]+.*$", npath1)) {
+      nunc <- sub("^(\\\\\\\\[^\\\\]+\\\\[^\\\\]+).*$", "\\1", npath1)
```

Comment:
I believe this is the backslash to end all other text, but short of the true name of ba'al, the soul-eater https://xkcd.com/1638/